worker/lease/doc.go

```diff
@@ -0,0 +1,22 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// Package lease is a worker that manages leases, which are used by individual
+// juju workers.
+// The manager works by using these operations:
+//   - claim
+//   - revoke
+//   - pin
+//   - query
+//   - block
+//
+// Workers will claim a lease, and they are either attributed or blocked. The
+// lease manager will keep track of all the blocked claims. When a lease
+// expires or gets revoked, then one of the claims of other workers will be
+// unblocked, thus assigning the lease to that worker.
+// Pin is a special case for making sure that a series upgrade doesn't change
+// the lease of an application. This means that the pinned lease will not
+// expire and the application will not lose its leadership. After the series
+// upgrade finishes it will be refreshed (extending its expiry date) and
+// finally unpinned, bringing operation back to normal.
```

Comment:
I'd try to roll this information into the list above, following Jordan's suggestion. Here it feels hard to parse because it's out of context. Also, it would be good to describe the other items in the list as well.